src/spatio_temporal_voxel_layer.cpp

```diff
@@ -662,24 +662,28 @@ void SpatioTemporalVoxelLayer::updateCosts( \
 }
 
 /*****************************************************************************/
-void SpatioTemporalVoxelLayer::UpdateROSCostmap(double* min_x, double* min_y, \
-                                                double* max_x, double* max_y)
+void SpatioTemporalVoxelLayer::UpdateROSCostmap( \
+                        double* min_x, double* min_y, \
+                        double* max_x, double* max_y, \
+                        std::unordered_set<volume_grid::occupany_cell>& updated_cells)
 /*****************************************************************************/
 {
   // grabs map of occupied cells from grid and adds to costmap_
   Costmap2D::resetMaps();
 
-  std::unordered_map<volume_grid::occupany_cell, uint>::iterator it;
-  for (it = _voxel_grid->GetFlattenedCostmap()->begin();
-       it != _voxel_grid->GetFlattenedCostmap()->end(); ++it)
+  auto cost_map = _voxel_grid->GetFlattenedCostmap();
+
+  for (const auto& cell: updated_cells)
```

Comment:
This entirely breaks the clearing logic and is much slower. You're only checking the cleared cells, not the newly marked cells. What you should do here is revert this entire function to what it was and add a new `for` loop to `touch` the `cleared_cells` set. The marked cells set is implicitly covered by iterating through the entire flattened costmap. 

---

If you wanted to be very fancy, the "best" thing to do here would be to change this logic so that `costmap_` wasn't being cleared and re-populated each iteration. So you store the diffs of the clears and the diffs of the marks each iterations and you updated the `costmap_` by those. That would remove the entire flattened costmap iteration here but you'd need to store a `marked_cells` set too. I'm not advocating for you to do this, but that would be one way to do this.

Reply:
Maybe I misunderstood how the marking logic is implemented, but it seems to me that obstacles are marked in the ROS costmap always 1 cycle after they are detected. *(I was going to create an issue about this, but wanted to focus on this PR before that).*

> The marked cells set is implicitly covered by iterating through the entire flattened costmap.

I agree with this. However, the flattened costmap is not up to date, but rather 1 cycle behind.

Obstacle information is only added to the flattened costmap in one place - inside the `SpatioTemporalVoxelGrid::PopulateCostmapAndPointcloud` function:

https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_grid.cpp#L256-L266

And from what I understood, the call stack for that function is:

1. https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_grid.cpp#L235
2. https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_grid.cpp#L159
3. https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_layer.cpp#L722

Meaning obstacle information is added to the costmap before marking actually happens:

https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_layer.cpp#L719-L743

When you call `SpatioTemporalVoxelGrid::Mark` (which in turn calls `SpatioTemporalVoxelGrid::operator()`), obstacle information is **stored in the voxel grid:**

https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_grid.cpp#L317

However, it is **not stored in the flattened costmap** - this only happen during the next cycle, when `SpatioTemporalVoxelGrid::ClearFrustums` is called.

**TL;DR:** obstacle information is added to the costmap based on marking information from the previous cycle. In other words, the obstacle information of the costmap is always lagging 1 cycle behind.

---

That is why the variable `updated_cells` holds information about both clearing and marking. Information about cleared cells is stored when calling `SpatioTemporalVoxelGrid::ClearGridPoint`, and information about marked cells is stored when calling `SpatioTemporalVoxelGrid::PopulateCostmapAndPointcloud`. For both cases, this happens inside the function `SpatioTemporalVoxelGrid::TemporalClearAndGenerateCostmap`.

Which is why I believe this PR doesn't actually break the current logic - it keeps it exactly the same. When you make the check `it->second >= _mark_threshold`, you are checking the flattened costmap that does not contain the most recent marking information - meaning it is the same information I stored in the `updated_cells` variable and thus we can avoid iterating the entire thing. 

https://github.com/SteveMacenski/spatio_temporal_voxel_layer/blob/e077795c89a280ffffa83db61af540464e34a892/src/spatio_temporal_voxel_layer.cpp#L672-L684

Reply:
> TL;DR: obstacle information is added to the costmap based on marking information from the previous cycle. In other words, the obstacle information of the costmap is always lagging 1 cycle behind.

You are correct, I recall making that design choice way back when but looking at it now, it seems like the easy solution is just to call `PopulateCostmapAndPointcloud` inside of `operator()` so that the `Mark` is adding to the cost map / grid points structures. Then you wouldn't have to change any of this.

Reply:
Wouldn't the refactoring be better suited for a different PR? I think it would be cleaner to keep things separate: this PR fixes the issue of the incorrect clearing, and another PR refactors the marking logic.

Reply:
What refactor? You just call a function and its done, with no changes needed to the UpdateROSCostmap function. this seems to refactor far more than what I'm suggesting

Reply:
I meant more about splitting PRs by their intent. The intent of this PR was to fix the clearing issue, so IMO fixing the marking issue should be a different PR.

Anyway, fixed the marking too. Created a new overload for the function `SpatioTemporalVoxelGrid::PopulateCostmapAndPointcloud` that receives a world pose.